### PR TITLE
metricbeat/module/redis: Add TLS and username support for Redis module

### PR DESCRIPTION
**Pull request by shmsr**
## What does this PR do?

PR adds support for TLS to Redis module in metricbeat. It also adds supports for accepting username (supported by by Redis 6.0+) from URI's userinfo or query parameter.

It has also has some nitpick changes pointed out by linters, etc.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- TLS support
- Username support
- Testing with Redis 5.x or lower
- Testing with Redis 6.x or higher

## How to test this PR locally

- Use a redis-server with TLS enabled
- Use a redis-server with username set (Redis 6.0+ should be used)
- Use a redis-server with password set and without username
- Use a redis-server with both TLS and AUTH (username and password set) enabled.

But how will get the certificate, CA, keys, etc.? 

- Invoke the script: https://github.com/redis/redis/blob/unstable/utils/gen-test-certs.sh (after cloning redis/redis)
- It'll generate the following:
  - tests/tls/ca.{crt,key}          Self signed CA certificate.
  - tests/tls/redis.{crt,key}       A certificate with no key usage/policy restrictions.
  - tests/tls/client.{crt,key}      A certificate restricted for SSL client usage.
  - tests/tls/server.{crt,key}      A certificate restricted for SSL server usage.
  - tests/tls/redis.dh              DH Params file.
- Now run the redis-server in TLS mode (example):
```
$ redis-server --tls-port 6378 --port 0 --tls-cert-file ./tests/tls/redis.crt --tls-key-file ./tests/tls/redis.key --tls-ca-cert-file ./tests/tls/ca.crt
```
- If you want to use redis-cli or any other client (metricbeat's redis module) then use the same `ca.crt`, `redis.crt` and `redis.key`:
```
redis-cli -p 6378 --tls --cert ./tests/tls/redis.crt --key ./tests/tls/redis.key --cacert ./tests/tls/ca.crt
```

## Related issues

- Fixes https://github.com/elastic/integrations/issues/5427

### CODE REVIEW

**Comment by shmsr** [comment]
Not sure how it added the reviewers automatically as I didn't and I just updated the project. So I've manually removed the assigned reviewers for now. But please free to review the changes as getting more eyeballs for the same is good.

**Comment by lalit-satapathy** [comment]
Ping: @ishleenk17

File: metricbeat/module/redis/redis.go [hunk 1]
```diff
@@ -208,27 +208,23 @@ func (p *Pool) DBNumber() int {
 }
 
 // CreatePool creates a redis connection pool
-func CreatePool(
-	host, password, network string,
-	dbNumber int,
-	maxConn int,
-	idleTimeout, connTimeout time.Duration,
-) *Pool {
+func CreatePool(host, username, password string, dbNumber int, config *ModuleConfig, connTimeout time.Duration) *Pool {
```
Comment (gpop63) [inline metricbeat/module/redis/redis.go:211]: The Dial func could be refactored to accept the `*tls.Config` and append the two additional TLS related options if it's enabled, what do you think?

```suggestion
func CreatePool(host, username, password string, dbNumber int, config *ModuleConfig, connTimeout time.Duration, tlsConfig *tls.Config) *Pool {
```

```go
		Dial: func() (rd.Conn, error) {
			dialOptions := []rd.DialOption{
				rd.DialUsername(username),
				rd.DialPassword(password),
				rd.DialDatabase(dbNumber),
				rd.DialConnectTimeout(connTimeout),
				rd.DialReadTimeout(connTimeout),
				rd.DialWriteTimeout(connTimeout),
			}

			if config.TLS.IsEnabled() {
				dialOptions = append(dialOptions, rd.DialUseTLS(true))
				dialOptions = append(dialOptions, rd.DialTLSConfig(tlsConfig))
			}

			return rd.Dial(config.Network, host, dialOptions...)
		},
```

Comment (shmsr) [reply metricbeat/module/redis/redis.go:211]: Good point. As I was using an extra field `UseTLS`, I kept it like this. But I think your way is more explicit and would be easier for readers who are trying to understand the code. I'll make this change.

File: metricbeat/module/redis/metricset.go [hunk 2]
```diff
@@ -35,32 +37,50 @@ type MetricSet struct {
...
 }
 
+type ModuleConfig struct {
+	IdleTimeout time.Duration     `config:"idle_timeout"`
+	Network     string            `config:"network"`
+	MaxConn     int               `config:"maxconn" validate:"min=1"`
+	TLS         *tlscommon.Config `config:"ssl"`
+
+	UseTLS       bool
+	UseTLSConfig *tls.Config
```
[the remark this answers is not in the document]
Comment (shmsr) [reply metricbeat/module/redis/metricset.go:47]: No. But as you even pointed out in  https://github.com/elastic/beats/pull/35240/files#r1218444763 that we can do it even without adding a new field.

**Comment by ishleenk17** [comment]
@shmsr 

General recommendation: 

It will be good to have only the enhancement related changes in this PR. 
The other generic syntax change etc changes can be part of a separate PR. 
It makes reviewing of the actual changes easier.

**Comment by shmsr** [comment]
> @shmsr
> 
> General recommendation:
> 
> It will be good to have only the enhancement related changes in this PR. The other generic syntax change etc changes can be part of a separate PR. It makes reviewing of the actual changes easier.

Yes, I agree and that's why I left out many such cosmetic changes. The changes that are non-related are mostly raised by golangci-lint ran as part of this repository's CI itself. If I don't fix them, the CI will fail because actions that runs the linter will fail.

File: metricbeat/module/redis/metricset.go [hunk 3]
```diff
@@ -35,32 +37,50 @@ type MetricSet struct {
...
+}
```
[the remark this answers is not in the document]
Comment (shmsr) [reply metricbeat/module/redis/metricset.go]: You are right. I checked other modules. They do have config in `config.go` and also use `DefaultConfig` function to return the default configuration. I've made the necessary change.

Thanks!

File: metricbeat/module/redis/redis.go (hunk 1 above)
Comment (shmsr) [reply metricbeat/module/redis/redis.go]: Done!

File: metricbeat/module/redis/metricset.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shmsr) [reply metricbeat/module/redis/metricset.go]: Now the config generation to is similar to other modules.

**Comment by shmsr** [comment]
> @shmsr please add a changelog entry in `CHANGELOG.next.asciidoc`

@gpop63 Done.

**Comment by ishleenk17** [comment]
Changes look good. Has this change been tested with TLS-enabled redis TLS mode?

**Comment by shmsr** [comment]
> Changes look good. Has this change been tested with TLS-enabled redis TLS mode?

Yes. I have even given the steps on how to test this PR locally: https://github.com/elastic/beats/pull/35240#issue-1686583037.

**Review by ishleenk17 — APPROVED** [review]
Looks good!
Please resolve conflicts and merge

**Review by gpop63 — APPROVED** [review]

**Review by cmacknz — APPROVED** [review]

**Pull request merged**